[WIP]: Added action buttons in the table component (new)

gokhaleprachi opened the pull request:
Fixes #329.

**Changes proposed in this pull request:**
- Added ability to define action button columns in the table component
- Changed storybook to show usage

**Newly added dependencies with [Bundlephobia](https://bundlephobia.com/) links:**
- N/A



<a href="https://gitpod.io/#https://github.com/HospitalRun/components/pull/369"><img src="https://gitpod.io/api/apps/github/pbs/github.com/gokhaleprachi/components.git/aa0fbf9a4e088bb8d1204a095e6c34fb5bf267fe.svg" /></a>

gokhaleprachi committed:
feat(table component): added action buttons, in separate columns for now

Added action buttons to the Table Component, each button currently is in its own column. Need to get
all buttons into a single column named actions before PR.

gokhaleprachi committed:
feat(table component): added action buttons,separate columns, w/o Header

Action buttons added to the table component, the buttons are each in thier own column.

feature #329, Ability to define actions for a row in the table

gokhaleprachi committed:
feat(table component): removed column def from react-table-config.d.ts

removed column def from react-table-config.d.ts as it was causing build issues

fix #329 in components

irvelervel commented:
Hi @gokhaleprachi , thank you for your work! Just a few suggestions for your code:
- we will probably never have the need to sort and filter the records by the buttons, so I think it makes sense to always set the `disableSorting` and `disableFiltering` properties to `false`, so you don't have to specify those parameters in every new table action column
- for the same reasons I think the properties `filterPlaceholder` and `Filter` are not necessary
- `action` as the type is more appropriate than `actions` imho because every other type is singular
- the `accessor` property should be always an empty string because we're not checking any data property
What do you think about these? Thank you again and have a nice day. :)

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

gokhaleprachi commented:
HI @irvelervel !

- I set `disableSorting` and `disableFiltering` to true for that reason, I thought that setting them to true would disable them. Correct me if I'm wrong!

- You are totally right that I don't need  `filterPlaceholder` and `Filter` in `action`. I'll get rid of those (and I will change the naming to `action` instead of `actions`)

- in terms of the accessor property for `action`, I believe it needs to be there as that is how we will access the custom onClick() that will change based the type of button the column is holding/ the data in the row (edit patient for example, will need to be a different link for each patient)

I'll go in and make the changes as soon as I get a confirmation that this all makes sense to you!

irvelervel commented:
Oh I'm sorry, I meant setting `disableSorting` and `disableFiltering` to `true` would disable them, just as you said. But from what I can see you are taking those values from the `columns` array in the table definition:
```
disableSorting: element.disableSorting,
disableFiltering: element.disableFiltering,
```
Because we are never going to actually set those properties to `false` in buttons, I think it makes more sense disabling them directly in the `helper.tsx` setting those lines to:
```
disableSorting: true,
disableFiltering: true,
```
so in the columns definition we don't have to specify those properties to `true` every time.

Regarding the last point, I think we need to shift the logic a little: in my opinion the callbacks for the buttons shouldn't be part of the data array, because it's not an info regarding the patients (and also it involves a lot of code repetition!). Why not setting the callback inside the column definition instead? So every button in every row will fire a set function.

fox1t commented on stories/table.stories.tsx:
```diff
@@ -20,6 +23,12 @@ storiesOf('Table', module)
         status: 'married',
         admin: true,
         date: new Date().toString(),
+        editOnClick: () => {
```
I don't like mixing functions and data. Can we find a better solution?

fox1t commented:
Hi @gokhaleprachi thanks for your contribution!

irvelervel commented on stories/table.stories.tsx:
```diff
@@ -20,6 +23,12 @@ storiesOf('Table', module)
         status: 'married',
         admin: true,
         date: new Date().toString(),
+        editOnClick: () => {
+          Toast('success', 'Edit button clicked!!', 'Success')
+        },
+        deleteOnClick: () => {
+          Toast('success', 'Delete button clicked!!', 'Success')
+        },
```
This code is better suited for the columns definition, so you don't have to redeclare it on every patient (and also mixes data and logic)

gokhaleprachi committed:
feat(table component): moved onclick logic for actions to columns

moved the button onclick logic from the data[] to columns[] for action columns. other small styling
fixes included.

#369

gokhaleprachi committed:
Merge branch 'tableComponentActions' of github.com:gokhaleprachi/components into tableComponentActions

jackcmeyer commented on src/components/Table/helper.tsx:
```diff
@@ -176,6 +177,27 @@ export const generateColumns = (columns: CustomColumn[]) => {
+          Cell: ({ row }: CellProps<object>) => (
+            <>
+              {element.id && (
+                <Button onClick={() => element.onClick(get(row.original, 'id', ''))}>
```
The onClick function should just return the entire object the row is based on, rather than the `id`. 

This makes the code more flexible and allows the client to decide what the want to do with the data.

jackcmeyer reviewed (changes requested):
For the most part looks good: 

Just some other things I have noticed: 

- [ ] storybook example does not have columnheaders above the table for the button columns. We need the ability to support column headers.
- [ ] I don't see the ability to change the color of the button anywhere

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

Commit:
Merge branch 'master' into tableComponentActions

tehKapa commented:
@gokhaleprachi do you need help?

kumikokashii commented:
This PR is for the ticket https://github.com/HospitalRun/components/issues/329, which is covered by another ticket, whose PR is https://github.com/HospitalRun/components/pull/419
